presto-docs/src/main/sphinx/connector/hive-cache.rst

```diff
@@ -0,0 +1,76 @@
+===================
+Hive Object Caching
+===================
+
+Querying object storage with the Hive connector is a very common use case for
+Presto, that often involves the transfer of large amounts of data. The objects
+are retrieved from HDFS, or any other supported object storage, by multiple
+workers and processed on these workers.
+
+Repeated queries with different parameters, or even different queries from
+different users, often access, and therefore transfer, the same objects. The
+Hive object caching can provide significant performance benefits, by avoiding
+the repeated network transfers and instead accessing copies of the objects from
+a local cache.
+
+.. warning::
+  Hive object caching is currently available as beta release only.
+
+.. contents::
+  :local:
+
+Architecture
+------------
+
+The Hive object caching provides a read-through cache. After first retrieval
+from storage by any query, objects are cached in the local cache storage on the
+workers. Objects are cached on local storage of each worker and managed by a
+bookkeeper component. The cache chunks are 1MB in size and are well suited for
+ORC or Parquet format objects.
+
+Configuration
+-------------
+
+The caching feature is part of the :doc:`Hive connector <./hive>` and can be
+activated in the catalog properties file:
+
+.. code-block:: none
+
+    connector.name=hive-hadoop2
+    hive.cache.enabled=true
+    hive.cache.location=/opt/hive-cache
+
+The cache operates on the coordinator and all workers accessing the object
+storage. The used networking ports for the managing bookkeeper and the data
```

Comment:
Not sure how much this is trying to abstract away Rubix being the implementation but I was sort of surprised that additional ports were needed for local caching. Does this mean Presto launches Rubix as an external process and then communicates with it via those ports? I'm not sure if linking to some Rubix reference documentation is appropriate here but would be good to know more about how those ports are used.

Other areas of interest:

1. Cache invalidation -  generally I don't think people tend to overwrite keys with new objects but does the implementation safeguard against this? Probably okay that it doesn't just would call it out.
2. Cache lifetime - I think people tend to do immutable deploys but if you restart a worker would it still be able to cache hit off those locally stored segments or does it effectively start over?

Reply:
I see, looking at #3420 it looks like there's an embedded mode and an externalized mode, ports only applying to the latter. May be good to touch on those in some way.

Reply:
@dain will have to chime in on these technical details.

Reply:
ports do apply to embedded mode too. The main different in embedded mode is that rubix starts as threads inside the presto jvm and in many places the thrift calls are avoided when communication has to be happen to local rubix servers. Rest of the details remain the same.